src/Streamly/Internal/Data/Parser/ParserD.hs

```diff
@@ -442,18 +425,19 @@ takeWhile1 predicate (Fold fstep finitial fextract) =
     initial = return Nothing
 
     step Nothing a =
-        if predicate a
-        then do
-            res <- finitial
-            case res of
-                FL.Partial s -> do
-                    sr <- fstep s a
-                    return
-                        $ case sr of
-                              FL.Partial r -> Partial 0 (Just r)
-                              FL.Done b -> Done 0 b
-                FL.Done b -> return $ Done 1 b
-        else return $ Error "takeWhile1: predicate failed on first element"
+        let err = Error "takeWhile1: predicate failed on first element"
+         in if predicate a
+            then do
+                res <- finitial
+                case res of
+                    FL.Partial s -> do
+                        sr <- fstep s a
+                        return
+                            $ case sr of
+                                  FL.Partial s1 -> Partial 0 (Just s1)
+                                  FL.Done b -> Done 0 b
+                    FL.Done _ -> return err
```

Comment:
When `initial` return a done the assumption is the fold succeeded? Why is the behavior of the fold taken into consideration? I can understand that this is `takeWhile*1*` but is this the correct behavior?

Reply:
The semantics of takeWhile1 are such that it has to consume at least 1 element or return error. It will be complicated to explain in the docs if you have a special case for that rule. Ideally we should use an accumulator here. But we do not have that type. Other option is to error out if a fold returning Done is passed.